### PR TITLE
Crypto-21: Integrate Get Coins API

### PR DESCRIPTION
**Pull request by vitalyaliskevych**
### Whats News
- I Integrated Get Coins API
### Issues
- [Crypto-21](https://trello.com/c/iVozW1QM)
### Video

https://user-images.githubusercontent.com/98754902/224101899-ee73e820-4d3d-4240-b8f4-40da555be111.mov

### CODE REVIEW

File: Crypto/Flows/Screens/IntroScreen/IntroViewModel.swift [hunk 1]
```diff
@@ -18,7 +18,7 @@ class IntroViewModel: ObservableObject {
     @Published var buttonIsShown = false
     
     var onResult: ((Result) -> Void)?
-    var coin = Coin(name: "", price: 1.2, icon: Image.btcIcn)
+//    var coin = Coins(coins: Coin())
```
Comment (oleksandr-syurpita) [inline Crypto/Flows/Screens/IntroScreen/IntroViewModel.swift:21]: remove if don't need

File: Crypto/Flows/Screens/MainScreen/MainLoading.swift [hunk 2]
```diff
@@ -8,19 +8,18 @@
...
+            VStack {
+                ScrollView {
+                    coinList
+                        .listRowBackground(Color.mainColor)
+                }
+                settingsBlock
             }
-            settingsBlock
+            .background(Color.mainColor)
+            .createToolbarMainView(text: "Crypto").navigationBarTitleDisplayMode(.inline)
```
Comment (oleksandr-syurpita) [inline Crypto/Flows/Screens/MainScreen/MainLoading.swift:21]: navigationBar next line

File: Crypto/Flows/Screens/MainScreen/MainView.swift [hunk 3]
```diff
@@ -9,18 +9,26 @@ import SwiftUI
...
-                coinList
-                    .listRowBackground(Color.mainColor)
+        ZStack {
+            if viewModel.loading {
+                MainLoadingView()
+            } else {
+                VStack {
+                    ScrollView {
+                        coinList
+                            .listRowBackground(Color.mainColor)
```
Comment (oleksandr-syurpita) [inline Crypto/Flows/Screens/MainScreen/MainView.swift:22]: can we pull this `listRowBackground` in `coinList` extension

File: Crypto/Flows/Screens/MainScreen/MainViewModel.swift [hunk 4]
```diff
@@ -6,32 +6,54 @@
...
 import Foundation
+import Combine
+import SwiftUI
 
 class MainViewModel: ObservableObject {
     
     enum Result {
         case onSettingsSelected
-        case onCoinItemSelected(coinName: Coin)
+        case onCoinItemSelected(coinName: Coins.Coin)
```
Comment (oleksandr-syurpita) [inline Crypto/Flows/Screens/MainScreen/MainViewModel.swift:16]: rename `coinName`

File: Crypto/Flows/Screens/MainScreen/MainViewModel.swift [hunk 5]
```diff
@@ -6,32 +6,54 @@
...
+    }
+    
+    func getCoins() {
+        self.loading = true
+        coinService
+            .getCoins()
+            .sink { [weak self] completion in
+                guard case .failure(let error) = completion else { return }
+                self?.output = error.localizedDescription
+                self?.loading = false
```
Comment (oleksandr-syurpita) [inline Crypto/Flows/Screens/MainScreen/MainViewModel.swift:52]: why need this here ?

File: Crypto/Services/CoinService.swift [hunk 6]
```diff
@@ -7,27 +7,43 @@
 
 import Foundation
 import SwiftUI
+import Combine
 
-class CoinService: ObservableObject {
+protocol CoinService {
+    func getCoins() -> AnyPublisher<Coins, Error>
```
Comment (oleksandr-syurpita) [inline Crypto/Services/CoinService.swift:13]: add empty line before func

File: Crypto/Services/Models/CoinsDTO.swift [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+//
+//  Created by Developer on 01.03.2023.
+//
+
+import Foundation
+import SwiftUI
+
+struct CoinsDTO: Codable {
+    
+    let coins: [CoinDTO]
```
Comment (oleksandr-syurpita) [inline Crypto/Services/Models/CoinsDTO.swift:13]: add empty line

File: Crypto/Services/Models/CoinsDTO.swift [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+        let imageURL: String
+    }
+}
+
+extension CoinsDTO: DomainModelConvertible {
+    
+    func toDomainModel() -> Coins {
+        Coins(coins: self.coins.map({
+            $0.toDomainModel()
+        }))
```
Comment (oleksandr-syurpita) [inline Crypto/Services/Models/CoinsDTO.swift:28]: bad alignment

File: Crypto/Utilities/Models/Coin.swift [hunk 9]
```diff
@@ -7,10 +7,16 @@
...
+    struct Coin: Identifiable {
+        
+        let id: String
+        let name: String
+        let price: Double
+        let imageURL: String
+    }
 }
 
+
```
Comment (oleksandr-syurpita) [inline Crypto/Utilities/Models/Coin.swift:21]: remove empty line

File: Crypto/Utilities/Models/Coin.swift [hunk 10]
```diff
@@ -7,10 +7,16 @@
...
+
```
Comment (oleksandr-syurpita) [inline Crypto/Utilities/Models/Coin.swift:22]: same

File: Crypto/Flows/Coordinator/MainCoordinator/MainCoordinator.swift [hunk 11]
```diff
@@ -17,16 +17,17 @@ class MainCoordinator: ObservableObject {
...
     
-    init(mainViewModel: MainViewModel) {
+    
+    init(mainViewModel: MainViewModelImpl) {
         self.mainViewModel = mainViewModel
         mainViewModel.onResult = { [weak self] result in
             switch result {
             case .onSettingsSelected:
                 self?.selectSettings()
             case .onCoinItemSelected(let coinName):
```
[the remark this answers is not in the document]
Comment (vitalyaliskevych) [reply Crypto/Flows/Coordinator/MainCoordinator/MainCoordinator.swift:29]: no

**Pull request merged**